### DISCUSSION TITLE
Auto update global variable that's dependant on another global variable

### DISCUSSION THREAD

**Discussion by anjanesh** [alpinejs/alpine#3700] [1. Help] [answered]
How do I update the a variable stored in the global state which is dependent on another variable ?

https://codepen.io/anjanesh/pen/qBQwyME
```html
<!DOCTYPE html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />    
    <title>AlpineJS auto update of variable values</title>
</head>
<body>
<script>
document.addEventListener('alpine:init', () =>
{
    Alpine.store('global_data',
    {
        products:
        [
            {
                name: 'batmobile',
                price_USD: 20,
                price_INR: 1660, // 20 x 83
            },
            {
                name: 'herbie',
                price_USD: 35,
                price_INR: 2905, // 35 x 83
            },
            {
                name: 'delorean',
                price_USD: 51,
                price_INR: 4233, // 51 x 83
            },
        ]
    });
});
</script>

<div x-data="">
    <template x-for="(product, index) in $store.global_data.products">
        <div>
            <span x-text="product.name"></span> : 
            <input x-model="product.price_USD" /> USD
            (INR : <span x-text="product.price_INR"></span>)
        </div>
    </template>
</div>

<script defer src="https://unpkg.com/alpinejs@3.12.3/dist/cdn.min.js"></script>
</body>
</html>
```
I don't want to do a @change on the input element because I don't want the input event handlers replicated everywhere.

**Comment by ekwoka** [comment]
Like you want it to auto convert?

use getters, or just a function even.

You can refactor the store to be like

```js
products: [
  {
    name: 'herbie',
    USD: 35,
    get INR() {
      return this.USD * 85
    }
  }
]
```

And access it as you already do.

or just make a function

```js
function USDtoINR (num) {
  return num * 85
}
```

and in then do
```html
<span x-text="USDtoINR(product.USD)"></span>
```

For your exact example, the second may be better.

But the getter can be better for others

**Reply by anjanesh** [reply]
I've updated my code : https://codepen.io/anjanesh/pen/qBQwyME

I need it to be bound both ways.

```html
<div x-data="">
    <template x-for="(product, index) in $store.global_data.products">
        <div>
            <span x-text="product.name"></span> : 
            <input x-model="product.price_USD" /> USD
            =
            <input x-model="product.price_INR" /> INR
        </div>
    </template>
</div>
```

**Reply by ekwoka** [reply]
use a getter/setter pair https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set

**Reply by anjanesh** [reply]
How do I "watch" an array's value ? `x-init="$watch('$store.global_data.products', value => console.log(value))"`

**Reply by ekwoka** [reply]
Why do you need to?

`$watch` is very rarely the write tool for the job.

**Reply by anjanesh** [reply]
If it can detect which array's variable got changed, then I can update the value of the corresponding array variable.

**Reply by ekwoka** [reply]
Why would you need to?

To update the converted price?

Just use a getter/setter pair 
get https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get
set https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set

**Reply by anjanesh** [reply]
So I got this done by having this in `Alpine.store('global_data')` :
```
set priceInINR(p)
{
    const product = this.products.find(product => product.name === p.name);
    product.price_INR = p.price_USD * 83;
},
```
and setting this in the click handler :
```
<input
    x-model="product.price_USD"
    x-on:input="$store.global_data.priceInINR = product"
/>
```
But I had to put `x-on:input="$store.global_data.priceInINR = {name: product.name, amount: product.price_USD}"` in all 3 input text fields.

Is there a way to have `$store.global_data.priceInINR` triggered automatically when `product.price_USD` changes ?

**Reply by ekwoka** [reply]
This is a bad way to do this...

Lots of extra work.

The product itself should have prices as getter setters.

Here's you're using a setter as a strange way to call a function which doesn't help

on the product

```js
{
   name: 'bobs burgers',
   price_USD: 100,
   set price_INR(val) {
     this.price_USD = val / 83
   },
   get price_INR() {
     return this.price_USD * 83
   },
 }
 ```
 
 Now you can just update it
 
 ```js
 product.price_USD = 15
console.log(product.price_INR)
```

**Reply by anjanesh** [reply]
But I have a predefined array of product objects. How do I add a setter to all of them ?
```javascript
products.forEach((product) => {
    Object.defineProperty(product, 'price_INR', {
        set: function (newPriceINR) {
            this.price_USD = newPriceINR / 83;
        },
    });
});
```
?

**Reply by ekwoka** [reply, marked as answer]
Even easier

Wherever you injest the products

```js
const INRPriceConverter = {
   set(val) {
     this.price_USD = val / 83
   },
   get() {
     return this.price_USD * 83
   }
}

const addINRPrice = (prod) => Object.defineProperty(product, 'price_INR', INRPriceCnverter)

this.products = getProducts().map(addINRPrice)
```

**Reply by anjanesh** [reply]
Thanks You for your response @ekwoka - I did not realize this kind of use-case of `set` and `get`. I kept trying `this.price_INR = val * 83;` and `return this.price_INR / 83` which gave "Maximum call stack size exceeded" until realized what your snippet really does. Thanks again.

**Reply by ekwoka** [reply]
Also, feel free to flip it to have the USD one be the computed ones. I am not sure for your spec which you'd prefer as the source of truth.

And nothing special would need to be done with the object if you just wanted to JSON stringify it to post to the backend :) It'll have both fields in sync.
